### PR TITLE
codegen: use preserve_static_offset intrinsic

### PR DESCRIPTION
**Pull request by amscanne**
Currently argument access uses pointer conversions with manual integer
arithmetic. This means that using the same values will often result in
LLVM choosing to save these intermediate values, which violates the
rules imposed by the verified if these are based on the context.

This situation is increasingly likely to happen as programs get larger.

Instead, we change to use proper pointer arithmetic and protect context
accesses using the `preserve_static_offsets` instrinsic, which was
introduced for Clang to prevent the same set of optimizations.

This results in optimized IR being generated that no longer uses manual
pointer casts and integer arithmetic, and ends up being converted to the
`llvm.bpf.getelementptr.and.load` intrinsic:

```
-  %14 = ptrtoint ptr %0 to i64
-  %15 = add i64 %14, 16
-  %16 = inttoptr i64 %15 to ptr
-  %17 = load volatile i64, ptr %16, align 8
+  %14 = call i64 (ptr, i1, i8, i8, i8, i1, ...) @llvm.bpf.getelementptr.and.load.i64(ptr elementtype(i8) %0, i1 true, i8 0, i8 1, i8 3, i1 false, i64 immarg 16)
```

The types used during IR generation are still very confusing, because
pointers are often treated as integer types. This could be significant
improved in the future, and types coerced only at the point where it is
needed.

##### Checklist

- [X] Language changes are updated in `man/adoc/bpftrace.adoc`
- [X] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: src/ast/passes/codegen_llvm.cpp [hunk 1]
```diff
@@ -4221,11 +4255,9 @@ void CodegenLLVM::probereadDatastructElem(Value *src_data,
   } else {
     // Read data onto stack
     if (data_type.IsCtxAccess() || data_type.is_btftype) {
+      // Types have already been suitable casted; just do the access.
```
Comment (jordalgo) [inline src/ast/passes/codegen_llvm.cpp:4258]:
```suggestion
      // Types have already been suitably casted; just do the access.
```

File: src/ast/passes/codegen_llvm.cpp [hunk 2]
```diff
@@ -4197,9 +4211,29 @@ void CodegenLLVM::probereadDatastructElem(Value *src_data,
                                           location loc,
                                           const std::string &temp_name)
 {
-  Value *src = b_.CreateAdd(src_data, offset);
+  // FIXME(amscanne): Pointers are often casted into integer types, and
+  // therefore we need to handle this case. Unfortunately, this means that we
+  // need to rewrap in the intrinsic to ensure that these continue to preserve
+  // offsets during access.  The solution to this to have have the local types
```
Comment (jordalgo) [inline src/ast/passes/codegen_llvm.cpp:4217]:
```suggestion
  // offsets during access.  The solution to this is to have the local types
```

File: src/ast/passes/codegen_llvm.cpp [hunk 3]
```diff
@@ -4197,9 +4211,29 @@ void CodegenLLVM::probereadDatastructElem(Value *src_data,
...
+  // accurately reflect the true types, rather than using `codegen_types`, and
+  // simply apply appropriate conversions at the moment needed.
```
Comment (jordalgo) [inline src/ast/passes/codegen_llvm.cpp:4219]: Not sure if I understand what this means. What is a 'codegen_type' ?

File: src/ast/passes/codegen_llvm.cpp [hunk 4]
```diff
@@ -2126,17 +2125,29 @@ void CodegenLLVM::visit(FieldAccess &acc)
...
-          raw = b_.CreateLoad(field_type,
-                              b_.CreateIntToPtr(src,
-                                                field_type->getPointerTo()),
-                              true);
-        else {
+        if (type.IsCtxAccess()) {
+          // We preserve pointer semantics. See `probereadDatastructElem` where
+          // the logic is essentially constrained in the same way.
+          if (expr_->getType() != b_.GET_PTR_TY()) {
+            if (!expr_->getType()->isPointerTy())
```
Comment (jordalgo) [inline src/ast/passes/codegen_llvm.cpp:2135]: I'm confused here - what's the difference between this if check and the one above `if (expr_->getType() != b_.GET_PTR_TY()) {`

**Comment by jordalgo** [comment]
Is it worth having a runtime test using a similar repro as in the issue: https://github.com/bpftrace/bpftrace/issues/3603

File: src/ast/passes/codegen_llvm.cpp (hunk 3 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.cpp:4219]: Ah, the `emit_codegen_types` parameter for `GetType`, which coerces pointer types into `Int64` by default (https://github.com/bpftrace/bpftrace/blob/master/src/ast/irbuilderbpf.cpp#L335).

I dropped the code that coerces the context immediately into a `Int64` as a builtin (and all tests are passing, so hopefully this is fine), but unfortunately that integer cast is still going to happen with any kind of variable assignment, e.g.

```
BEGIN {
  @test = 1;
}

tracepoint:syscalls:sys_enter_kill {
  $a = args;
  print($a.pid);
  exit();
}
```

Therefore it necessary to coerce back to a pointer and rewrap in the intrinsic.

File: src/ast/passes/codegen_llvm.cpp (hunk 4 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.cpp:2135]: If the type is not the correct pointer type, then it is either a) an integer, or b) a pointer of another type. Coercing into the correct pointer type is one of two different operations, depending on which is the case.

File: src/ast/passes/codegen_llvm.cpp (hunk 2 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.cpp]: Fixed.

File: src/ast/passes/codegen_llvm.cpp (hunk 1 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.cpp]: Fixed.

**Comment by amscanne** [comment]
> Is it worth having a runtime test using a similar repro as in the issue: #3603 ?

I did add a codegen test that is similar to the reproducer (https://github.com/bpftrace/bpftrace/pull/3629/files#diff-2ef6096262f7a2a7c30b2ba1e4e4057e2dcfe77cd954109da249c143f795146e), but it's probably not that helpful given that it affects a *lot* of generated code (fundamentally moving from the integer add op to the getelementptr op). Given the coverage, I could just convert this to a runtime test? Seems like it might be more likely to catch regressions if the codegen tests are updated w/o paying attention to generated code (which obviously, no one would do).

File: src/ast/passes/codegen_llvm.h [hunk 5]
```diff
@@ -327,6 +327,11 @@ class CodegenLLVM : public Visitor {
     return module_->getDataLayout().getTypeAllocSize(s);
   }
 
+  Value *getContext()
+  {
+    return b_.preserveStaticOffset(ctx_);
```
Comment (viktormalik) [inline src/ast/passes/codegen_llvm.h:332]: Could we wrap `ctx_` in `preserve_static_offsets` when assigning to it? That should lead to less code changes and prevent future regressions by accidentally using `ctx_` instead of `getContext()`.

File: src/ast/passes/codegen_llvm.cpp (hunk 3 above)
Comment (viktormalik) [reply src/ast/passes/codegen_llvm.cpp:4219]: The question is whether we should completely drop this "store pointers as integers" thing. I'm not sure if has any benefits at all. I know it's probably out of the scope of this PR but it would allow to simplify things here. I'm just wondering how much work it would be to convert everything to use raw pointers and not ints.

File: src/ast/irbuilderbpf.cpp [hunk 6]
```diff
@@ -2479,12 +2479,22 @@ CallInst *IRBuilderBPF::CreateSkbOutput(Value *skb,
...
+Value *IRBuilderBPF::CreateKFuncArg(Value *ctx,
+                                    SizedType &type,
+                                    std::string &name)
 {
   assert(type.IsIntTy() || type.IsPtrTy());
+
+#if LLVM_VERSION_MAJOR <= 14
+  // Older versions of LLVM check that the type matches the GEP. Newer versions
+  // do not use this cast because we need to specifically preserve the pattern
+  // `GEP(preserveStaticOffsets(ctx))`, which doesn't exist on older versions.
```
Comment (viktormalik) [inline src/ast/irbuilderbpf.cpp:2491]: So the pointer cast would drop the `preserve_static_offsets` attribute off `ctx`, right?

What if we used `llvm.bpf.getelementptr.and.load` directly? Wouldn't that simplify the code as we wouldn't need these ifdefs? I'm just curious, not sure if that's even possible.

**Review by viktormalik — COMMENTED** [review]
Looks good to me overall, just some high-level comments to clarify some things.

Also, it would be nice to reference some resource for the `preserve_static_offset` feature, both from the commit message and from the code. Either the commit https://github.com/llvm/llvm-project/commit/030b8cb1561db4161b108b59044717d89026cf70 or the [Phabricator link](https://reviews.llvm.org/D133361).

File: src/ast/passes/codegen_llvm.cpp (hunk 4 above)
Comment (ajor) [reply src/ast/passes/codegen_llvm.cpp:2135]: Does the pointer being the wrong type only apply for LLVM < 15? For newer LLVM versions there's only one pointer type, right?

If that's right, could you wrap it in a `#if LLVM_VERSION_MAJOR <= 14` check please? That'll make it easier to find and remove when we eventually deprecate the older LLVM versions.

File: src/ast/passes/codegen_llvm.cpp (hunk 3 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.cpp]: I'm strongly in favor of that, but I think that's a bigger exercise (and would probably be much easier with some broader cleanup first, to make it easier to reason about the types). I've refactored this change to try to reduce the scope a bit, as I'm looking for a relatively narrow fix. (Happy to push separately on the broader bits.)

File: src/ast/passes/codegen_llvm.cpp (hunk 4 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.cpp]: I've refactored this change a bit (might need to iterate to ensure it's working well for all versions), which I think makes this a lot easier to understand and centralizes all these conversions. It keeps the LLVM14-specific stuff, but they are generally less scattered anyways.

**Review by viktormalik — COMMENTED** [review]
Just a couple of nits and one test-related change suggestion.

File: tests/codegen/preserve_static_offsets.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#include "common.h"
+
+namespace bpftrace {
+namespace test {
+namespace codegen {
+
+TEST(codegen, preserve_static_offsets)
```
[the remark this answers is not in the document]
Comment (amscanne) [reply tests/codegen/preserve_static_offsets.cpp:7]: Agreed, @jordalgo also had this suggestion and I forgot to convert it. This is done now, it is a runtime regression test.

File: src/ast/irbuilderbpf.cpp [hunk 8]
```diff
@@ -2807,6 +2800,40 @@ llvm::Value *IRBuilderBPF::CreatePtrOffset(const SizedType &type,
   return CreateMul(index, getInt64(elem_size));
 }
 
+llvm::Value *IRBuilderBPF::SafeGEP(llvm::Type *ty,
```
[the remark this answers is not in the document]
Comment (amscanne) [reply src/ast/irbuilderbpf.cpp:2803]: Done.

File: src/ast/passes/codegen_llvm.h (hunk 5 above)
Comment (amscanne) [reply src/ast/passes/codegen_llvm.h]: No longer relevant, since I've move things around to make this a bit cleaner and eliminate this method. (I swear I had already responded to this chain, but it must have gotten lost?)

File: src/ast/irbuilderbpf.cpp [hunk 9]
```diff
@@ -2807,6 +2802,40 @@ llvm::Value *IRBuilderBPF::CreatePtrOffset(const SizedType &type,
   return CreateMul(index, getInt64(elem_size));
 }
 
+llvm::Value *IRBuilderBPF::CreateSafeGEP(llvm::Type *ty,
```
Comment (jordalgo) [inline src/ast/irbuilderbpf.cpp:2805]: Nice and clean with good comments. Thanks!

File: src/ast/irbuilderbpf.h [hunk 10]
```diff
@@ -266,6 +266,19 @@ class IRBuilderBPF : public IRBuilder<> {
                                llvm::Value *index,
                                AddrSpace as);
 
+  // Safely handle pointer references by wrapping the address with the
+  // intrinsic `preserve_static_offset` [1], which will ensure that LLVM does
+  // not apply certain basic optimizations (notably, saving any intermediate
+  // offset from this pointer). This is required for the context pointer, which
+  // cannot be modified. This method also automatically handles casts from
```
Comment (jordalgo) [inline src/ast/irbuilderbpf.h:273]: nit:

```suggestion
  // offset from this pointer). This is required for the context pointer, which,
  // if modified, will trigger an error in the verifier. This method also automatically handles casts from
```

File: tests/runtime/regression [hunk 11]
```diff
@@ -108,3 +108,12 @@ NAME unaligned key with aligned value
 PROG BEGIN { @mapA["aaaabbb", 0] = 1; @mapA["ccccdddd", 0] = 1; }
 EXPECT Attaching 1 probe...
 TIMEOUT 1
+
+# In some cases, LLVM can generate code that violates the BPF verifier
+# constraint that the context pointer is not modified. #3629 introduces the use
+# of intrinsics that should ensure that this doesn't happen, but this runtime
+# test exercises a pattern that previously produced code that failed to verify.
```
Comment (jordalgo) [inline tests/runtime/regression:115]: Nit: you could also add a comment to the original issue.

**Review by jordalgo — APPROVED** [review]
few nits but otherwise this looks great!

**Review by viktormalik — APPROVED** [review]
Nice!

File: tests/runtime/regression (hunk 11 above)
Comment (amscanne) [reply tests/runtime/regression]: Added link to both

**Comment by jordalgo** [comment]
One last thing to consider is a changelog entry. It's a pretty edge case bug at the moment and these changes are pretty deep into the internals so not sure it's worth it.

**Comment by amscanne** [comment]
> One last thing to consider is a changelog entry. It's a pretty edge case bug at the moment and these changes are pretty deep into the internals so not sure it's worth it.

Yeah, I'd just consider this to be part of the standard set of bug fixes, and would have a soft preference to avoid calling out that in the changelog.
